rust/hyperlane-core/src/types/mod.rs

```diff
@@ -20,6 +22,50 @@ mod message;
 /// 20-byte ids (e.g ethereum addresses)
 pub mod identifiers;
 
+/// Error indicating the internal type does not match the type it is being
+/// converted to
+#[derive(thiserror::Error)]
+pub struct InvalidType;
+
+impl Debug for InvalidType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self}")
+    }
+}
+
+impl Display for InvalidType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Incorrect internal type")
+    }
+}
+
+/// Quickly create a new primitive type wrapping enum
+macro_rules! define_primitive_hash_enum {
```

Comment:
Wanna flag that (iiuc) there's some similar logic that's included in https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2505 now that we'll be moving to some types defined within hyperlane-core

Because this PR isn't super critical I'd recommend making this PR to whatever branch https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/2505 ends up in (i.e. either wait for that PR to merge into `trevor/sealevel-validator-rebase` and make this against `trevor/sealevel-validator-rebase`, or if `trevor/sealevel-validator-rebase` is merged into main by then to put this PR against main)